### PR TITLE
Documentation: use @allowed_from to document allowed contexts

### LINKED ISSUE #21061: Document where APIs can be called from using doxygen
## Introduction

Right now we have no standard way for documenting what APIs can be called from where and if they for example are allowed to be called from thread context or ISR context or if they can be used during initialization of the system for example.

### Problem description

Documentation of the above is spread around the project either in the implementation itself or in the API docs or in the guides, but there is no 1 single place where we reliably and consistently document this.

### Proposed change

Use a custom tag in doxygen to document the context from which an API can be called from

## Detailed RFC

Add a new alias to doxygen (\@allowed_from) and use it consistently to document from what contexts APIs are allowed to be called.

### Proposed change (Detailed)

for example:
```diff
diff --git a/doc/zephyr.doxyfile.in b/doc/zephyr.doxyfile.in
index 147f94f629..087fc17892 100644
--- a/doc/zephyr.doxyfile.in
+++ b/doc/zephyr.doxyfile.in
@@ -224,6 +224,7 @@ ALIASES                = "rst=\verbatim embed:rst:leading-asterisk" \
                          "endrst=\endverbatim"

 ALIASES               += "req=\xrefitem req \"Requirement\" \"Requirements\" "
+ALIASES            += "allowed_from=\xrefitem context \"Allowed from\" \"Allowed from\" "

 # This tag can be used to specify a number of word-keyword mappings (TCL only).
 # A mapping has the form "name=value". For example adding "class=itcl::class"
```
will allow us to do this:

```diff
diff --git a/include/kernel.h b/include/kernel.h
index 23c0fd17f0..2ab916bcd4 100644
--- a/include/kernel.h
+++ b/include/kernel.h
@@ -1890,7 +1890,7 @@ __syscall void k_queue_init(struct k_queue *queue);
  * -EINTR and K_POLL_STATE_CANCELLED state (and per above, subsequent
  * k_queue_get() will return NULL).
  *
- * @note Can be called by ISRs.
+ * @allowed_from Threads, ISRs
  *
  * @param queue Address of the queue.
  *
```

which will result in:

image

instead of what we have now:

image

The above will also generate 1 single page with all the information:

image

### Dependencies

Doxygen changes

### Concerns and Unresolved Questions

How do we enforce this for example?

## Alternatives

Alternative is to create some table or adhoc documentation that will always be out of sync. If we make this a requirement for any new API, then we can always rely on the autogenerated doxygen documentation.

### PR DESCRIPTION
**Pull request by nashif**
This just transforms what we have right now in form of an unofficial @note to a
doxygen command the ultimitealy should be applied to all APIs to document the
context they can be called from.

A fix for a loose doxygen group in a test is also included.

Resolves #21061

### CODE REVIEW

**Review by pabigot — CHANGES_REQUESTED** [review]
There is active work on #18970 in draft form in #20959 to address concerns about API behavior and use across both kernel and non-kernel API, which I expect to have in reviewable form by early January.  At this time "Allowed from" does not adequately cover the attributes that must be identified (in particular it fails to identify reschedule points, which affect thread behavior).

I would appreciate an opportunity to complete that work and get it through the API review process before rushing to a solution that may not cover the distinctions we need.

**Comment by nashif** [comment]
> I would appreciate an opportunity to complete that work and get it through the API review process before rushing to a solution that may not cover the distinctions we need.

Sure, this can wait, however I am not sure exactly how to read #20959 and where all of this will end up, it would be great to have this in the form of a proposed change.

**Review by dbkinder — CHANGES_REQUESTED** [review]
These new doxygen comments (using @allowed_from to create an xrefitem) will only show up in the doxygen-generated HTML and not in the Sphinx/Breathe API documentation we currently publish.  Breathe ignores these xrefitem entries (but does properly handle @note items), so that's something to think about.  Depending on the resolution of the discussion started by @pabigot, we may want to instead use an alias to create @note directives for this kind of information so it can be handled by and appear in the breathe-processed API docs as well.

File: include/kernel.h [hunk 1]
```diff
@@ -776,6 +780,8 @@ extern FUNC_NORETURN void k_thread_user_mode_enter(k_thread_entry_t entry,
  * need to be.
  * Note that NULL shouldn't be passed as an argument.
  *
+ * @allowed_from Supervisor Mode
```
Comment (andrewboie) [inline include/kernel.h]: Everything is allowed from supervisor mode.
I think it's more useful to indicate which core kernel APIs are callable from user mode (although I suspect nearly all or all of them are tagged with __syscall)

File: doc/zephyr.doxyfile.in [hunk 2]
```diff
@@ -224,6 +224,7 @@ ALIASES                = "rst=\verbatim embed:rst:leading-asterisk" \
                          "endrst=\endverbatim"
 
 ALIASES 	       += "req=\xrefitem req \"Requirement\" \"Requirements\" "
+ALIASES            += "allowed_from=\xrefitem context \"Allowed from\" \"Allowed from\" "
```
Comment (ulfalizer) [inline doc/zephyr.doxyfile.in]: Why the random change in indentation?

Comment (ulfalizer) [reply doc/zephyr.doxyfile.in]: Looks like there's a random tab in there. Could get rid of it.

**Comment by nashif** [comment]
need another approach that supports rst

**Pull request closed without merging**